autoscheduler/scraper/tests/api_tests.py

```diff
@@ -0,0 +1,253 @@
+    def setUp(self):
+        self.client = APIClient()
+        self.courses = [
+            Course(id='123123', dept='CSCE', course_num='181',
+                   title='Introduction to Computing', term='201931', credit_hours=3),
+            Course(id='123124', dept='CSCE', course_num='315',
+                   title='Programming Studio', term='201931', credit_hours=3),
+            Course(id='123125', dept='COMM', course_num='203',
+                   title='Public Speaking', term='201831', credit_hours=3),
+            Course(id='123126', dept='COMM', course_num='203',
+                   title='Public Speaking', term='201931', credit_hours=3),
+            Course(id='123127', dept='LAW', course_num='7500S',
+                   title='Sports Law', term='202031', credit_hours=None),
+        ]
+        test_instructors = [
+            Instructor(id='Akash Tyagi'),
+            Instructor(id='John Moore'),
+        ]
+        for instructor in test_instructors:
+            instructor.save()
+        self.sections = [
+            Section(crn=12345, id='000001', subject='CSCE', course_num='310',
+                    section_num='501', term_code='201931', min_credits='3',
+                    honors_only=False, web_only=False, max_enrollment=50,
+                    current_enrollment=40, instructor=test_instructors[0]),
+            Section(crn=12346, id='000002', subject='CSCE', course_num='310',
+                    section_num='502', term_code='201931', min_credits='3',
+                    honors_only=False, web_only=False, max_enrollment=50,
+                    current_enrollment=40, instructor=test_instructors[1],
+                    instructor_gpa=3.2),
+        ]
+        self.meetings = [
+            Meeting(id='0000010', meeting_days=[True] * 7, start_time=time(11, 30),
+                    end_time=time(12, 20), meeting_type='LEC', section=self.sections[0]),
+            Meeting(id='0000011', meeting_days=[True] * 7, start_time=time(9, 10),
+                    end_time=time(10), meeting_type='LEC', section=self.sections[0]),
+            Meeting(id='0000020', meeting_days=[True] * 7, start_time=time(11, 30),
+                    end_time=time(12, 20), meeting_type='LEC', section=self.sections[1]),
+            Meeting(id='0000021', meeting_days=[False] * 7, start_time=time(9, 10),
+                    end_time=time(10), meeting_type='LAB', section=self.sections[1]),
+        ]
+        for course in self.courses:
+            course.save()
+        for section in self.sections:
+            section.save()
+        for meeting in self.meetings:
+            meeting.save()
+
+    def test_api_terms_displays_all_terms(self):
+        """ Tests that /api/terms returns ordered list of all terms in database """
+        # Arrange
+        expected = {
+            '201831': 'Fall 2018 - College Station',
+            '201931': 'Fall 2019 - College Station',
+            '202031': 'Fall 2020 - College Station',
+        }
```

Comment:
This should be something like

```python
expected = {
    '202011': 'Spring 2020 - College Station',
    '202012': 'Spring 2020 - Galveston',
    '202013': 'Spring 2020 - Qatar'
    '202014': 'Spring 2020 - Half Year Term'
    '201931': 'Fall 2019 - College Station'
    ....
}
```

Ideally this test (and the `/api/terms` route) should update to the most recent semesters that can be taken, but this should be fine for now.

See [here](https://github.com/aggie-coding-club/Automatic-Aggie-Scheduler/blob/ddae9f45dd00e742229b60cf73d27d6d1b270e2b/autoscheduler/scraper/banner_requests.py#L33) in `banner_requests.py` to see how term formatting works.

We may want to restrict it to be only College Station / Galveston / Qatar if that's what they choose frequently using cookies or whatever, but that's something we can do at a later point.

Reply:
On that same note, we may need some form of storage of the most recent term that's updated somehow so that the server automatically knows what the most recent / upcoming term is. More of a thought rather than a review comment though lol.

Reply:
> 
> 
> This should be something like
> 
> ```python
> expected = {
>     '202011': 'Spring 2020 - College Station',
>     '202012': 'Spring 2020 - Galveston',
>     '202013': 'Spring 2020 - Qatar'
>     '202014': 'Spring 2020 - Half Year Term'
>     '201931': 'Fall 2019 - College Station'
>     ....
> }
> ```

I think there should be separate tests making sure that term codes are formatted correctly, since it better follows the principle of only testing one component at a time, and we're just wanting to make sure that all the term codes in the database are displayed. If you disagree though, respond to this comment and I can change the test